vars/generateGoBenchmarkDiff.groovy

```diff
@@ -0,0 +1,84 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+
+Generate a Go benchmark report by comparing the existing benchmark with
+the `TARGET_BRANCH` variable if exists.
+
+  generateGoBenchmarkDiff(file: 'bench.out', filter: 'exclude')
+**/
+
+def call(Map args = [:]) {
+  if(!isUnix()){
+    error('generateGoBenchmarkDiff: windows is not supported yet.')
+  }
+  def file = args.containsKey('file') ? args.file : error('generateGoBenchmarkDiff: file parameter is required.')
+  def filter = args.get('filter', 'none')
+
+  if (!fileExists("${file}")) {
+    error('generateGoBenchmarkDiff: file does not exist.')
+  }
+
+  // Prepare the compare context with.
+  def compareWith = getCompareWithFileIfPossible(file: file, output: 'build')
+
+  // Run the report generation script
+  if (compareWith?.trim()) {
+    runBenchmarkDiff(filter: filter, file: file, compareWith: compareWith)
+  }
+}
+
+def getCompareWithFileIfPossible(Map args = [:]) {
+  def compareWith = ''
+    // This is only available for Pull Requests
+  if (env.CHANGE_TARGET) {
+    try {
+      dir("${args.output}/${env.CHANGE_TARGET}") {
+        projectName = env.JOB_NAME.replace(env.JOB_BASE_NAME, env.CHANGE_TARGET)
+        copyArtifacts(filter: "${args.file}", flatten: true, optional: true, projectName: projectName, selector: lastWithArtifacts())
+      }
+    } catch(e) {
+      log(level: 'INFO', text: 'generateGoBenchmarkDiff: it was not possible to copy the previous build.')
+    } finally {
+      compareWith = "${args.output}/${env.CHANGE_TARGET}/${args.file}"
+      if (!fileExists("${compareWith}")) {
+        compareWith = ''
+      }
+    }
+  }
+  return compareWith
+}
+
+def runBenchmarkDiff(Map args = [:]) {
+  def diffReport = getReportFileName()
+  def flags = ''
+  if (args.filter.equals('exclude')) {
+    flags = "| grep -v 'all equal' | grep -v '~'"
+  }
+  withGoEnv(pkgs: [ "golang.org/x/perf/cmd/..." ] ) {
+    sh(label: 'generateGoBenchmarkDiff', script: "benchstat ${args.file} ${args.compareWith} ${flags} | tee ${diffReport}")
```

Comment:
can we change the order of the arguments? in general `benchstat` is run with `benchstat old.result new.result` so that you get the right data in the right columns, otherwise we'll get the results backwards.

Reply:
@marclop , I'm a bit confused with the `grep` filters:

```
benchstat 7.17.out 8.3.out | grep -v 'all equal' | grep -v '~' # Exclude unchanged results. 
name                                                old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
```

for such I downloaded the bench files for 7.17 and 8.3:

```
wget https://apm-ci.elastic.co/job/apm-server/job/apm-server-mbp/job/7.17/301/artifact/src/github.com/elastic/apm-server/bench.out -O 7.17.out
wget https://apm-ci.elastic.co/job/apm-server/job/apm-server-mbp/job/8.3/32/artifact/src/github.com/elastic/apm-server/bench.out -O 8.3.out  

```

Is the above the expected report output?

Reply:
Do you mean `benchstat` should use the `bench.out` that was generated using those new env variables?

Reply:
yes, we should modify the way we run the benchmarks, so they're run a few times (5) is a good place to start and then `benchstat` should use the newly generated `bench.out`.

Reply:
I think we still need to modify `.ci/scripts/bench.sh` line 4 so it includes`BENCH_COUNT=5`. I didn't make that change in the PR I merged today, so that's still pending.

Reply:
https://github.com/elastic/apm-server/pull/8476 is the one with the proposal and [build 2](https://apm-ci.elastic.co/job/apm-server/job/apm-server-mbp/job/PR-8476/2/) is the one that will run the benchmark for the PR, let's see how it goes